proposals/NNNN-vk-const-id.md

```diff
@@ -0,0 +1,109 @@
+<!-- {% raw %} -->
+
+# SPIR-V Specialization Constants
+
+*   Proposal: [NNNN](NNNN-vk-const-id.md)
+*   Author(s): [Steven Perron](https://github.com/s-perron)
+*   Status: **Design In Progress**
+
+*During the review process, add the following fields as needed:*
+
+*   PRs:
+
+    [#143179](https://github.com/llvm/llvm-project/pull/143179)
+    [#143180](https://github.com/llvm/llvm-project/pull/143180)
+
+## Introduction
+
+SPIR-V has a feature called
+[Specialization Constants](https://registry.khronos.org/SPIR-V/specs/unified1/SPIRV.html#_specialization_2)
+that allows shader code to have constants that are set by the driver. This
+feature is exposed in HLSL via the `vk::constant_id` attribute.
+
+## Motivationus
+
+This feature is commonly used in Vulkan to avoid creating multiple versions of
+the same shader for each value of the constant. It was implemented in DXC, so we
+want to match that behavior.
+
+## Proposed solution
+
+The `vk::constant_id` attribute applies to a global constant scalar variable:
+
+```
+[[vk::constant_id(0)]] const float my_constant = 1.0f;
+```
+
+This indicates that the value of `my_constant` can be overridden by the driver
+by providing a value for specialization constant ID `0`. If no value is provided
+by the driver for this ID, then `my_constant` will take its default value of
+`1.0f`.
+
+Sema will modify the variable to be `static` (giving it internal linkage) and
+change its initializer to be a call to a new built-in
+`__builtin_get_spirv_spec_constant`. This built-in takes the specialization
+constant ID and the default value as parameters. The AST for this variable would
+be:
+
+```
+|-VarDecl 0x5595791c2960 <vk.spec-constant.hlsl:1:24, col:50> col:36 used my_constant 'const hlsl_private float' static cinit
+| |-CallExpr 0x5595791c2e58 <col:50> 'float'
+| | |-ImplicitCastExpr 0x5595791c2e40 <col:50> 'float (*)(unsigned int, float) noexcept' <FunctionToPointerDecay>
+| | | `-DeclRefExpr 0x5595791c2da8 <col:50> 'float (unsigned int, float) noexcept' lvalue Function 0x5595791c2b50 '__builtin_get_spirv_spec_constant_float' 'float (unsigned int, float) noexcept'
+| | |-ImplicitCastExpr 0x5595791c2e90 <col:3> 'unsigned int' <IntegralCast>
+| | | `-IntegerLiteral 0x5595791c2a60 <col:3> 'int' 0
+| | `-FloatingLiteral 0x5595791c2a40 <col:50> 'float' 1.000000e+00
+| `-HLSLVkConstantIdAttr 0x5595791c29c8 <col:3, col:20> 0
+```
+
+During code generation, this built-in will be replaced with a call to the
+existing SPIR-V builtin, `__spirv_SpecConstant`, which has the same semantics:
+
+```
+llvm @_ZL11my_constant = internal addrspace(10) global float 0.000000e+00, align 4
+...
+define internal spir_func void @__cxx_global_var_init() #4 {
+entry:
+  %0 = call token @llvm.experimental.convergence.entry()
+  %1 = call reassoc nnan ninf nsz arcp afn float @_Z20__spirv_SpecConstantif(i32 0, float 1.000000e+00)
```

Comment:
I don't think the `nnan`, `ninf`, `nsz` flags should be there.
As for `arcp` and `afn` not sure if those should be there without requiring fastmath no?

Reply:
I just copied what was being created by my implementation. I agree those should not be there, but I wonder how they got there. I'll try to figure that out, and update accordingly.

Reply:
This seems to be coming from a general option in the compiler. All function in the llvm-ir have the attributes `"no-infs-fp-math"="true" "no-nans-fp-math"="true" "no-signed-zeros-fp-math"="true"`. I'll remove them from the description here, so that it is not assumed that the must be there.

I won't change the compiler implementation. This is an orthongal issue.